Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp

```diff
@@ -602,6 +607,25 @@ static void webkitWebViewBaseSetToplevelOnScreenWindow(WebKitWebViewBase* webVie
         priv->pageProxy->activityStateDidChange(flagsToUpdate);
 }
 
+static void deviceScaleFactorChanged(WebKitWebViewBase* webkitWebViewBase)
+{
+#if GTK_CHECK_VERSION(4, 12, 0)
+    // Allow correctly handling fractional scaling values by directly getting the scale factor from Gdk.
+    GtkRoot* root = gtk_widget_get_root(GTK_WIDGET(webkitWebViewBase));
+    if (!root)
+        return;
+    GdkSurface* surface = gtk_native_get_surface(GTK_NATIVE(root));
+    if (!surface)
+        return;
```

Comment:
Why not  `gtk_native_get_surface(gtk_widget_get_native(GTK_WIDGET(webkitWebViewBase)));` ?